### ISSUE TITLE
Single Page Apps

### ISSUE THREAD

**Issue by benaadams** [Microsoft/ApplicationInsights-JS#2]
```trackPageView``` correctly uses ```performance.timing```, however for SPA using xhr this is always first page load time. Use of [Resource Timing](http://www.w3.org/TR/resource-timing/) can get round this.

PerformanceTiming 
```javascript
{connectEnd: 1429922346337
connectStart: 1429922346216
domComplete: 1429922378350
domContentLoadedEventEnd: 1429922378038
domContentLoadedEventStart: 1429922378028
domInteractive: 1429922378028
domLoading: 1429922377408
domainLookupEnd: 1429922346216
domainLookupStart: 1429922346216
fetchStart: 1429922346215
loadEventEnd: 1429922378361
loadEventStart: 1429922378350
navigationStart: 1429922345941
redirectEnd: 0
redirectStart: 0
requestStart: 1429922346337
responseEnd: 1429922377776
responseStart: 1429922377406
secureConnectionStart: 0
unloadEventEnd: 0
unloadEventStart: 0}
```
PerformanceResourceTiming
```javascript
{connectEnd: 0
connectStart: 0
domainLookupEnd: 0
domainLookupStart: 0
duration: 13486.342999996617
entryType: "resource"
fetchStart: 10071187.829000002
initiatorType: "xmlhttprequest"
name: "http://..../"
redirectEnd: 0
redirectStart: 0
requestStart: 0
responseEnd: 10084674.171999998
responseStart: 0
secureConnectionStart: 0
startTime: 10071187.829000002}
```

However that's also a little messy in that you need to search though it  to find your xhr page load and it has everything in it, images, xhr (including ```dc.services.visualstudio.com/v2/track```) etc so you need to pick out the correct one. Also you need to periodically clear the buffer when it gets full to add more recent stuff (e.g. 150 items in chrome):
```javascript
if (window.performance) {
    if (performance.clearResourceTimings) {
        performance.clearResourceTimings();
    } else if (performance.webkitClearResourceTimings) {
        performance.webkitClearResourceTimings()
    }
}
```
Equally, the user will need to manually calculate the domProcessing, perhaps using ```setImmedate``` or polyfill for it. So there isn't an immediate "this is the right way solution" that I can think of that can be automated, without bundling it into the xhr loader + page updater.

However... all the the properties of ```Telemetry.PageViewPerformance``` can be calculated so it would be useful to be able to pass in a ```PageViewPerformance``` either to ```trackPageView``` or a new function ```trackDataLoad``` e.g.

```typescript
public trackPageView(name?: string, url?: string, 
    properties?: Object, measurements?: Object, 
    timings?: PageViewPerformance) {
```
That it would use instead of ```new Telemetry.PageViewPerformance``` when provided?

**Comment by scsouthw** [comment]
You can use startTrackPageView(page name) and stopTrackPageView(page name) for SPA. This time between these calls will be reported as the page load duration.

On Apr 24, 2015 9:05 PM, Ben Adams <notifications@github.com> wrote:

trackPageView correctly uses performance.timing, however for SPA using xhr this is always first page load time. Use of Resource Timing<http://www.w3.org/TR/resource-timing/> can get round this.

PerformanceTiming

{connectEnd: 1429922346337
connectStart: 1429922346216
domComplete: 1429922378350
domContentLoadedEventEnd: 1429922378038
domContentLoadedEventStart: 1429922378028
domInteractive: 1429922378028
domLoading: 1429922377408
domainLookupEnd: 1429922346216
domainLookupStart: 1429922346216
fetchStart: 1429922346215
loadEventEnd: 1429922378361
loadEventStart: 1429922378350
navigationStart: 1429922345941
redirectEnd: 0
redirectStart: 0
requestStart: 1429922346337
responseEnd: 1429922377776
responseStart: 1429922377406
secureConnectionStart: 0
unloadEventEnd: 0
unloadEventStart: 0}

PerformanceResourceTiming

{connectEnd: 0
connectStart: 0
domainLookupEnd: 0
domainLookupStart: 0
duration: 13486.342999996617
entryType: "resource"
fetchStart: 10071187.829000002
initiatorType: "xmlhttprequest"
name: "http://..../"
redirectEnd: 0
redirectStart: 0
requestStart: 0
responseEnd: 10084674.171999998
responseStart: 0
secureConnectionStart: 0
startTime: 10071187.829000002}

However that's also a little messy in that you need to search though it to find your xhr page load and it has everything in it, images, xhr (including dc.services.visualstudio.com/v2/track) etc so you need to pick out the correct one. Also you need to periodically clear the buffer when it gets full to add more recent stuff (e.g. 150 items in chrome):

if (window.performance) {
    if (performance.clearResourceTimings) {
        performance.clearResourceTimings();
    } else if (performance.webkitClearResourceTimings) {
        performance.webkitClearResourceTimings()
    }
}

Equally, the user will need to manually calculate the domProcessing, perhaps using setImmedate or polyfill for it. So there isn't an immediate "this is the right way solution" that I can think of that can be automated, without bundling it into the xhr loader + page updater.

However... all the the properties of Telemetry.PageViewPerformance can be calculated so it would be useful to be able to pass in a PageViewPerformance either to trackPageView or a new function trackDataLoad e.g.

public trackPageView(name?: string, url?: string,
    properties?: Object, measurements?: Object,
    timings?: PageViewPerformance) {

That it would use instead of new Telemetry.PageViewPerformance when provided?

**Comment by benaadams** [comment]
Then you loose some of the timing richness as shown in the portal:

![capture](https://cloud.githubusercontent.com/assets/1142958/7331570/3399bf98-eb0e-11e4-9fcc-3f476cf7647a.PNG)

**Comment by benaadams** [comment]
It does then produce a more interesting time line:

![capture](https://cloud.githubusercontent.com/assets/1142958/7359715/9ee6fea0-ed39-11e4-839e-1de78520693a.PNG)

**Comment by scsouthw** [comment]
I see. Thanks for putting this together :)

We'll review and try to include this with the next version.

**Comment by MaxShehovtsov** [comment]
@benaadams I think what I'm missing is how exactly the images/static resources are going to be filtered out (or taken into calculation). I see the following problems with your approach:
- each SPA page can have multiple ajax calls - how do we sum it up?
- during SPA page load the app may do other ajax calls (not related to this SPA page) - how do we filter out irrelevant ajaxes
- other resources (images/css) - do we take them in or out?

Given these problems here's how we see 
The idea we currently have is that PageViewPerformance is really a descriptor for the first page load (as it's got DOM loading). For SPA's it make sense to only show the "page load time" which is defined by user for different SPA models - for some there might be no ajax calls at all, for instance.
Combined with tacking ajax calls separately and showing them for each SPA (like dependency calls for server requests) will complete it.

Does it make sense?

P.s. I'm sorry for not getting to the pull request for a long time.

**Comment by benaadams** [comment]
@MaxShehovtsov if I understand it currently calling ```trackPageView``` will put ```performance.timing``` into the envelope which doesn't make sense passed initialization? Might be easier if I run through a usage example?

For example in this 23hr period I have 1.12M page views and with 132.18 pages per session on average; the initialisation when it takes the hit and loads all the resources (js, css etc) is important; but isn't predominantly what I'm interested in; once you are on the site everything is controlled via hash paths and in place page replacement, meaning js and css is only loaded once; so in a traditional sense there will only be one ```performance.timing``` or page load per session and its only 0.75% of an average users page views.

![capture](https://cloud.githubusercontent.com/assets/1142958/8049087/e83e40c0-0e51-11e5-905b-28acec8cc92e.png)

However I want to know what's going on with the other pages which are pages in the practice sense; other than not causing a full browser refresh; so I do want to drill into them:

![capture2](https://cloud.githubusercontent.com/assets/1142958/8049174/73138f1a-0e53-11e5-88b0-a327fff7ac75.PNG)

Chrome looks a little anomalous to the other desktop browsers so I want to change the filter to look more closely at that specifically (at this point I think I broke it as everything changed to errors, but you get the idea)

Mostly I'm just asking to be able to provide timings as an optional parameter ( #3 old )  and then leave it as an exercise for the developer to derive this data and the right time to call it. e.g. is it a trackPageView or a trackEvent? Perhaps also be able to attach timings to events.

In terms of actually technically getting the timings from ```window.performance``` I prepared an example gist that shows how you can do it with jquery and its ajaxComplete https://gist.github.com/benaadams/f7d1e9136d8c60939467 though I wouldn't necessarily recommend it as it will intercept all jquery ajax requests, which may more may not be interesting; but its fairly easily adapted to something else; and its mostly about finding things in perforance and keeping it empty (as it stops recording when its full) However...

> each SPA page can have multiple ajax calls - how do we sum it up?

You can't without domain knowledge; hence the ability for the developer with domain knowledge to provide timings and collate them as necessary.

> during SPA page load the app may do other ajax calls (not related to this SPA page) - how do we filter out irrelevant ajaxes

As above

> other resources (images/css) - do we take them in or out?

Static resources are a different beast; but I suppose if someone wants to record them as events and attach timings they could...

**tl;dr** interested in being able to pass timings as a parameter to trackPageView; which overrides those set in ```window.performance.timing```

**Comment by benaadams** [comment]
Just looking at server side doesn't give a good picture as that says these requests are taking 20ms

![capture](https://cloud.githubusercontent.com/assets/1142958/8060084/2c9b97b8-0ebb-11e5-9fcf-54b8b32b069b.png)

**Comment by MaxShehovtsov** [comment]
No progress for a while (I guess unit tests were a road blocker). Closing the issue.
